server/controllers/thread.controller.js

```diff
@@ -52,12 +52,13 @@ function findLastReadMessageId(thread, allMessages) {
             continue; // eslint-disable-line no-continue
         }
         if (maxCreatedMessageId) {
-            if (maxCreatedDate < allMessages[j].readAt) {
-                maxCreatedDate = allMessages[j].readAt;
+            if (maxCreatedDate < allMessages[j].createdAt) {
```

Comment:
This still doesn't make sense to me. `maxCreatedDate` will be created as `null`, so the check at line 55 will always evaluate to the same thing. Additionally. `maxCreatedDate` is never returned or used. I'm pretty sure I'm missing something here that may need to be documented in-line.

Reply:
The first loop iteration falls through to the else clause where it is initialized because `maxCreatedMessageId` is also falsey to start.
`maxCreatedDate` is used in the line 55 comparison to see if the iteration message is newer than the current max message. `{ id: maxCreatedId, createdAt: maxCreatedDate}` is a subobject of the newest entry in the `allMessages` that has been seen at any point in the loop.
I'm more inclined to rewrite this without a for loop than to comment it if it is hard to read.

Reply:
Is changing to not iterate over every message a big lift?

Reply:
I think this could be written without the loop but i think it would require a second sql query.

Reply:
It was looping over fewer messages than I thought, so I left it. I simplified the implementation to make it more readable.
I think it is possible to compute these values in pure sql using some kind of subquery but I don't know enough to do it quickly